gradle-plugin-testing/src/main/java/com/palantir/gradle/plugintesting/DiscoveryTestClassesTask.java

```diff
@@ -0,0 +1,159 @@
+
+    private static final Logger log = LoggerFactory.getLogger(DiscoveryTestClassesTask.class);
+
+    @InputFiles
+    public abstract ConfigurableFileCollection getTestClasspath();
+
+    @InputFiles
+    public abstract ConfigurableFileCollection getTestSourceFiles();
+
+    @InputFiles
+    public abstract ConfigurableFileCollection getRuntimeClasspath();
+
+    @OutputFile
+    public abstract RegularFileProperty getOutputFile();
+
+    @OutputFile
+    public abstract RegularFileProperty getDiscoveredTestsFile();
+
+    @Inject
+    public abstract ProjectLayout getProjectLayout();
+
+    @Inject
+    public abstract ProviderFactory getProviderFactory();
+
+    @Input
+    public abstract Property<File> getParentPath();
+
+    @Input
+    public abstract ListProperty<String> getExtraArguments();
+
+    @Input
+    public abstract Property<String> getTestClassType();
+
+    public DiscoveryTestClassesTask() {
+        setClasspath(getProject().files(getRuntimeClasspath(), getTestClasspath()));
+        getDiscoveredTestsFile()
+                .convention(getProviderFactory()
+                        .zip(
+                                getTestClassType(),
+                                getProjectLayout().getBuildDirectory(),
+                                (type, buildDir) ->
+                                        buildDir.dir("tests-discovery").file(String.format("%s-test-classes", type))));
+        getOutputFile()
+                .convention(getProviderFactory()
+                        .zip(
+                                getTestClassType(),
+                                getProjectLayout().getBuildDirectory(),
+                                (type, buildDir) -> buildDir.dir("tests-discovery")
+                                        .file(String.format("%s-test-classes-paths", type))));
+        getMainClass().set("com.palantir.gradle.plugintesting.DiscoverTestsMain");
+        getArgumentProviders().add(this::getArguments);
+
+        doLast(new Action<Task>() {
+            @Override
+            public void execute(Task _task) {
+                try {
+                    Path parentPath = getParentPath().get().toPath();
+                    Set<String> discoveredTestClassesPaths =
+                            Files.readAllLines(getDiscoveredTestsFile()
+                                            .getAsFile()
+                                            .get()
+                                            .toPath())
+                                    .stream()
+                                    .map(this::toFilePath)
+                                    .collect(Collectors.toSet());
+                    Set<String> discoveredTestFiles = getTestSourceFiles().getAsFileTree().getFiles().stream()
+                            .map(File::toPath)
+                            .filter(path -> pathMatches(path, discoveredTestClassesPaths))
+                            .map(path -> parentPath.relativize(path).toString())
```

Comment:
is the idea here that you'll get a relative path relative to the root directory then? if so, we should call it `rootPath` instead of parent path.